File: _experimental/_ocaml/words/lib/words.ml
let reverse_string str =
  let len_str = String.length str in
  if len_str <= 1
  then str
  else
    let buff = Buffer.create len_str in
    for i = len_str - 1 downto 0 do
      Buffer.add_char buff str.[i]
    done;
    Buffer.contents buff

let is_palindrome str =
  let cleaned_str = Str.global_replace (Str.regexp "[ \t\n\r\x0C]+") "" str in
  let rec loop i j =
    if i >= j then true
    else if cleaned_str.[i] <> cleaned_str.[j] then false
    else loop (i + 1) (j - 1)
  in
  loop 0 (String.length cleaned_str - 1)

let compute_lps_array (patt : string) : int array =
  let len_patt = String.length patt in
  let lps = Array.make len_patt 0 in
  let rec lps_loop i len =
    if i >= len_patt then ()
    else
      if patt.[i] = patt.[len] then
        begin
          lps.(i) <- len + 1;
          lps_loop (i + 1) (len + 1)
        end
      else if len > 0 then
          lps_loop i (len - 1)
      else
        lps_loop (i + 1) len in
  lps_loop 1 0;
  lps

let kmp_search sub str =
  let len_sub = String.length sub in
  let len_str = String.length str in
  if len_sub = 0
  then true
  else if len_sub > len_str
  then false
  else
    let lps = compute_lps_array sub in
    let rec kmp_loop i j =
      if j = len_sub then true
      else if i >= len_str then false
      else if str.[i] = sub.[j] then kmp_loop (i + 1) (j + 1)
      else if j > 0 then kmp_loop i lps.(j - 1)
      else kmp_loop (i + 1) j
    in
    kmp_loop 0 0

let lcs str1 str2 =
  let len1 = String.length str1 in
  let len2 = String.length str2 in
  let dp = Array.make_matrix (len1 + 1) (len2 + 1) 0 in
  for i = 1 to len1 do
    for j = 1 to len2 do
      if str1.[i - 1] == str2.[j - 1]
      then dp.(i).(j) <- dp.(i - 1).(j - 1) + 1
      else dp.(i).(j) <- max dp.(i - 1).(j) dp.(i).(j - 1)
    done;
  done;
  dp.(len1).(len2)

let max a b =
  if a > b then a
  else b

File: _experimental/_erlang/words/src/words.erl
-module(words).

-export([reverse_string/1, is_palindrome/1, kmp_search/2, lcs/2]).

reverse_string(Str) ->
  reverse_helper(Str, []).

reverse_helper([], Buffer) ->
  Buffer;
reverse_helper([H | T], Buffer) ->
  reverse_helper(T, [H | Buffer]).

is_palindrome(Str) ->
  CleanedStr = re:replace(Str, "\\s+", "", [global, {return, list}]),
  CleanedStr == reverse_string(CleanedStr).

compute_lps_array(Str) ->
  LenStr = length(Str),
  Lps = array:new(LenStr, [{default, 0}]),
  update_lps(1, 0, Lps, Str, LenStr).

update_lps(I, Len, Lps, Str, LenStr) ->
  if
    I >= LenStr ->
      Lps;
    true ->
      StrBinary = list_to_binary(Str),
      CharI = binary:part(StrBinary, I, 1),
      CharLen = binary:part(StrBinary, Len, 1),
      if
        CharI == CharLen ->
          LpsUpdated = array:set(I, Len + 1, Lps),
          update_lps(I + 1, Len + 1, LpsUpdated, Str, LenStr);
        true ->
          if
            Len > 0 ->
              PrevLen = array:get(Len - 1, Lps),
              update_lps(I, PrevLen, Lps, Str, LenStr);
            true ->
              LpsUpdated = array:set(I, 0, Lps),
              update_lps(I + 1, Len, LpsUpdated, Str, LenStr)
          end
      end
  end.

kmp_search(Sub, Str) ->
  LenSub = length(Sub),
  LenStr = length(Str),
  if 
    LenSub == 0 -> 
      true;
    LenSub > LenStr -> 
      false;
    true ->
      Lps = compute_lps_array(Sub),
      kmp_loop(0, 0, Sub, Str, LenSub, LenStr, Lps)
  end.

kmp_loop(I, J, Sub, Str, LenSub, LenStr, Lps) ->
  if
    J == LenSub ->
      true;
    I >= LenStr ->
      false;
    true ->
      StrBinary = list_to_binary(Str),
      SubBinary = list_to_binary(Sub),
      CharStr = binary:part(StrBinary, I, 1),
      CharSub = binary:part(SubBinary, J, 1),
      if
        CharStr == CharSub ->
          kmp_loop(I + 1, J + 1, Sub, Str, LenSub, LenStr, Lps);
        true ->
          if
            J > 0 ->
              NewJ = array:get(J - 1, Lps),
              kmp_loop(I, NewJ, Sub, Str, LenSub, LenStr, Lps);
            true ->
              kmp_loop(I + 1, J, Sub, Str, LenSub, LenStr, Lps)
          end
      end
  end.

lcs(Str1, Str2) ->
  Len1 = length(Str1),
  Len2 = length(Str2),
  if
    (Len1 == 0) orelse (Len2 == 0) -> 
      0;
    true ->
      Size = (Len1 + 1) * (Len2 + 1),
      Dp = array:new(Size, [{default, 0}]),
      FinalDp = for_i(1, Str1, Str2, Len1, Len2, Dp),
      Index = Len1 * (Len2 + 1) + Len2,
      array:get(Index, FinalDp)
  end.

for_i(I, Str1, Str2, Len1, Len2, Dp) ->
  if
    I > Len1 ->
      Dp;
    true ->
      DpUpdated = for_j(I, 1, Str1, Str2, Len1, Len2, Dp),
      for_i(I + 1, Str1, Str2, Len1, Len2, DpUpdated)
  end.

for_j(I, J, Str1, Str2, Len1, Len2, Dp)->
  if
    J > Len2 ->
      Dp;
    true ->
      Str1Binary = list_to_binary(Str1),
      Str2Binary = list_to_binary(Str2),
      Char1 = binary:part(Str1Binary, I - 1, 1),
      Char2 = binary:part(Str2Binary, J - 1, 1),
      IndexCurr = I * (Len2 + 1) + J,
      IndexPrevRowCol = (I - 1) * (Len2 + 1) + (J - 1),
      IndexPrevRow = (I - 1) * (Len2 + 1) + J,
      IndexPrevCol = I * (Len2 + 1) + (J - 1),
      if
        Char1 == Char2 ->
          MatchVal = array:get(IndexPrevRowCol, Dp),
          DpUpdated = array:set(IndexCurr, MatchVal + 1, Dp),
          for_j(I, J + 1, Str1, Str2, Len1, Len2, DpUpdated);
        true ->
          Val1 = array:get(IndexPrevRow, Dp),
          Val2 = array:get(IndexPrevCol, Dp),
          MaxVal = max(Val1, Val2),
          DpUpdated = array:set(IndexCurr, MaxVal, Dp),
          for_j(I, J + 1, Str1, Str2, Len1, Len2, DpUpdated)
      end
  end.
